src/omnicore/fees.cpp

```diff
@@ -0,0 +1,478 @@
+        for (uint32_t itPropertyId = startPropertyId; itPropertyId < _my_sps->peekNextSPID(ecosystem); itPropertyId++) {
+            int64_t distributionThreshold = getTotalTokens(itPropertyId) / OMNI_FEE_THRESHOLD;
+            distributionThresholds[itPropertyId] = distributionThreshold;
+        }
+    }
+
+}
+
+// Gets the current amount of the fee cache for a property
+int64_t COmniFeeCache::GetCachedAmount(const uint32_t &propertyId)
+{
+    assert(pdb);
+    // Get the fee history, set is sorted by block so last entry is most recent
+    std::set<feeCacheItem> sCacheHistoryItems = GetCacheHistory(propertyId);
+    if (!sCacheHistoryItems.empty()) {
+        std::set<feeCacheItem>::iterator endIt = sCacheHistoryItems.end();
+        --endIt;
+        feeCacheItem mostRecentItem = *endIt;
+        return mostRecentItem.second;
+    } else {
+        return 0; // property has never generated a fee
+    }
+}
+
+// Zeros a property in the fee cache
+void COmniFeeCache::ClearCache(const uint32_t &propertyId, int block)
+{
+    if (msc_debug_fees) PrintToLog("ClearCache starting (block %d, property ID %d)...\n", block, propertyId);
+    const std::string key = strprintf("%010d", propertyId);
+    std::set<feeCacheItem> sCacheHistoryItems = GetCacheHistory(propertyId);
+    if (msc_debug_fees) PrintToLog("   Iterating cache history (%d items)...\n",sCacheHistoryItems.size());
+    std::string newValue;
+    if (!sCacheHistoryItems.empty()) {
+        for (std::set<feeCacheItem>::iterator it = sCacheHistoryItems.begin(); it != sCacheHistoryItems.end(); it++) {
+            feeCacheItem tempItem = *it;
+            if (tempItem.first == block) continue;
+            if (!newValue.empty()) newValue += ",";
+            newValue += strprintf("%d:%d", tempItem.first, tempItem.second);
+            if (msc_debug_fees) PrintToLog("      Readding entry: block %d amount %d\n", tempItem.first, tempItem.second);
+        }
+        if (!newValue.empty()) newValue += ",";
+    }
+    if (msc_debug_fees) PrintToLog("   Adding zero valued entry: block %d\n", block);
+    newValue += strprintf("%d:%d", block, 0);
+    leveldb::Status status = pdb->Put(writeoptions, key, newValue);
+    assert(status.ok());
+    ++nWritten;
+
+    PruneCache(propertyId, block);
+
+    if (msc_debug_fees) PrintToLog("Cleared cache for property %d block %d [%s]\n", propertyId, block, status.ToString());
+}
+
+// Adds a fee to the cache (eg on a completed trade)
+void COmniFeeCache::AddFee(const uint32_t &propertyId, int block, const int64_t &amount)
+{
+    if (msc_debug_fees) PrintToLog("Starting AddFee for prop %d (block %d amount %d)...\n", propertyId, block, amount);
+
+    // Get current cached fee
+    int64_t currentCachedAmount = GetCachedAmount(propertyId);
+    if (msc_debug_fees) PrintToLog("   Current cached amount %d\n", currentCachedAmount);
+
+    // Add new fee and rewrite record
+    if ((currentCachedAmount > 0) && (amount > INT64_MAX - currentCachedAmount)) {
```

Comment:
Should be `std::numeric_limits<int64_t>::max()` instead of `INT64_MAX`. This needs `#include <limits>`.